jwst/stpipe/config.py

```diff
@@ -0,0 +1,217 @@
+
+        Parameters
+        ----------
+        asdf_file : asdf.AsdfFile
+
+        Returns
+        -------
+        stpipe.config.StepConfig
+
+        Raises
+        ------
+        asdf.ValidationError
+            If the file does not validate against any of the
+            recognized schemas.
+        """
+        try:
+            _validate_asdf(asdf_file, _CONFIG_SCHEMA_URI)
+            return cls._from_tree(asdf_file.tree)
+        except asdf.ValidationError as e:
+            # Maybe it's an old-style ASDF file:
+            try:
+                _validate_asdf(asdf_file, _LEGACY_CONFIG_SCHEMA_URI)
+                return cls._from_legacy_tree(asdf_file.tree)
+            except asdf.ValidationError:
+                # Raise the original error so that we encourage
+                # use of the new config file format.
+                raise e
+
+    @classmethod
+    def _from_tree(cls, tree):
+        class_name = tree["class"]
+        name = tree["name"]
+        parameters = deepcopy(tree["parameters"])
+        steps = [cls._from_tree(s) for s in tree.get("steps", [])]
+        return StepConfig(class_name, name, parameters, steps)
+
+    @classmethod
+    def _from_legacy_tree(cls, tree):
+        def _inner(parameters, name):
+            class_name = parameters.pop("class", None)
+            steps = [_inner(s, n) for n, s in parameters.pop("steps", {}).items()]
+            return StepConfig(class_name, name, parameters, steps)
+
+        parameters = deepcopy(tree["parameters"])
+        name = parameters.pop("name")
+        return _inner(parameters, name)
+
+
+def export_config(step):
+    """
+    Export a step's current parameters to a StepConfig object.
+
+    Parameters
+    ----------
+    step : stpipe.Step
+
+    Returns
+    -------
+    stpipe.config.StepConfig
+    """
+    class_name = get_fully_qualified_class_name(step)
+    name = step.name
+
+    parameters = step.get_pars()
+    # The Pipeline class includes step parameters, but we're
+    # going to collect those ourselves so we can also get
+    # ahold of the name and class name.
+    parameters.pop("steps", None)
+
+    steps = [
+        export_config(getattr(step, step_name))
+        for step_name, _ in getattr(step, "step_defs", {}).items()
+    ]
+
+    return StepConfig(class_name, name, parameters, steps)
+
+
+def _validate_asdf(asdf_file, schema_uri):
+    """
+    Validate an ASDF file against the config schema.
+
+    Parameters
+    ----------
+    asdf_file : asdf.AsdfFile
+        Open config file.
+    schema_uri : str
+        StepConfig schema URI (must be registered with the ASDF entry point).
+
+    Raises
+    ------
+    asdf.ValidationError
+    """
+    # TODO: We should add a method on AsdfFile to facilitate this:
+    schema = asdf.schema.load_schema(schema_uri, asdf_file.resolver)
+    tagged_tree = asdf.yamlutil.custom_tree_to_tagged_tree(asdf_file.tree, asdf_file)
+    asdf.schema.validate(tagged_tree, asdf_file, schema)
```

Comment:
If one loads this using the `custom_schema` kwarg when creating the `AsdfFile` instance, one can just call `af.validate()` to do the validation.

Reply:
That's true.  But this module didn't necessarily load the file and doesn't have control over what `custom_schema` was specified.